src/mame/heathkit/tlb.cpp

```diff
@@ -0,0 +1,635 @@
+DEFINE_DEVICE_TYPE(SUPER19, heath_super19_tlb_device, "heath_super19_tlb", "Heath Terminal Logic Board w/Super19 ROM");
+DEFINE_DEVICE_TYPE(WATZ, heath_watz_tlb_device, "heath_watz_tlb", "Heath Terminal Logic Board w/Watzman ROM");
+DEFINE_DEVICE_TYPE(ULTRA, heath_ultra_tlb_device, "heath_ultra_tlb", "Heath Terminal Logic Board w/Ultra ROM");
+
+heath_tlb_device::heath_tlb_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+		: heath_tlb_device(mconfig, TLB, tag, owner, clock)
+{
+}
+
+heath_tlb_device::heath_tlb_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
+		: device_t(mconfig, type, tag, owner, clock)
+		, write_sd(*this)
+		, m_palette(*this, "palette")
+		, m_maincpu(*this, "maincpu")
+		, m_crtc(*this, "crtc")
+		, m_ace(*this, "ins8250")
+		, m_beep(*this, "beeper")
+		, m_p_videoram(*this, "videoram")
+		, m_p_chargen(*this, "chargen")
+		, m_mm5740(*this, "mm5740")
+		, m_kbdrom(*this, "keyboard")
+		, m_kbspecial(*this, "MODIFIERS")
+{
+}
+
+TIMER_CALLBACK_MEMBER(heath_tlb_device::key_click_off)
+{
+	m_keyclickactive = false;
+
+	if (!m_keyclickactive && !m_bellactive)
+		m_beep->set_state(0);
+}
+
+TIMER_CALLBACK_MEMBER(heath_tlb_device::bell_off)
+{
+	m_bellactive = false;
+
+	if (!m_keyclickactive && !m_bellactive)
+		m_beep->set_state(0);
+}
+
+void heath_tlb_device::mem_map(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0000, 0x1fff).mirror(0x2000).rom();
+	map(0x4000, 0x4100).mirror(0x3e00).ram();
+	map(0xc000, 0xc7ff).mirror(0x3800).ram().share("videoram");
+}
+
+void heath_tlb_device::io_map(address_map &map)
+{
+	map.unmap_value_high();
+	map.global_mask(0xff);
+	map(0x00, 0x00).mirror(0x1f).portr("SW401");
+	map(0x20, 0x20).mirror(0x1f).portr("SW402");
+	map(0x40, 0x47).mirror(0x18).rw(m_ace, FUNC(ins8250_device::ins8250_r), FUNC(ins8250_device::ins8250_w));
+	map(0x60, 0x60).mirror(0x1E).w(m_crtc, FUNC(mc6845_device::address_w));
+	map(0x61, 0x61).mirror(0x1E).rw(m_crtc, FUNC(mc6845_device::register_r), FUNC(mc6845_device::register_w));
+	map(0x80, 0x80).mirror(0x1f).r(FUNC(heath_tlb_device::kbd_key_r));
+	map(0xa0, 0xa0).mirror(0x1f).r(FUNC(heath_tlb_device::kbd_flags_r));
+	map(0xc0, 0xc0).mirror(0x1f).w(FUNC(heath_tlb_device::key_click_w));
+	map(0xe0, 0xe0).mirror(0x1f).w(FUNC(heath_tlb_device::bell_w));
+}
+
+
+// Keyboard encoder masks
+#define KB_ENCODER_KEY_VALUE_MASK 0x7f
+#define KB_ENCODER_CONTROL_KEY_MASK 0x80
+
+// Keyboard flag masks
+#define KB_STATUS_SHIFT_KEYS_MASK 0x01
+#define KB_STATUS_CAPS_LOCK_MASK  0x02
+#define KB_STATUS_BREAK_KEY_MASK  0x04
+#define KB_STATUS_ONLINE_KEY_MASK 0x08
+#define KB_STATUS_REPEAT_KEYS_MASK 0x40
+#define KB_STATUS_KEYBOARD_STROBE_MASK 0x80
+
+void heath_tlb_device::device_start()
+{
+
+	m_key_click_timer = timer_alloc(FUNC(heath_tlb_device::key_click_off), this);
+	m_bell_timer = timer_alloc(FUNC(heath_tlb_device::bell_off), this);
+}
```

Comment:
The two members previously saved (when this code was in h19.cpp), where `m_keyclickactive` and `m_bellactive`. These are very transient values get reset to false after a timer goes off(6 - 200 mSec). If the state was saved with the value set to true, since the timer state was not being saved, until a new key was press, `m_keyclickactive` would remain true, and until a Ctrl-G was press `m_bellactive` would remain true. During testing, since these variables were not being initialized to false, when starting the emulator after the first keypress, the bell would stay on until I pressed Ctrl-G. When I added the fix to initialize both to false, I looked at the saving, and didn't feel it was needed. Let me know if you think it needs to be saved.

Reply:
The timer state is automatically saved.